src/main/scala/mesosphere/marathon/core/task/Task.scala

```diff
@@ -83,39 +83,48 @@ object Task {
 
     private[this] def hasStartedRunning: Boolean = status.startedAt.isDefined
 
+    //scalastyle:off cyclomatic.complexity method.length
     override def update(update: TaskStateOp): TaskStateChange = update match {
       // case 1: now running
-      case TaskStateOp.MesosUpdate(MarathonTaskStatus.Running(mesosStatus), now) if !hasStartedRunning =>
+      case TaskStateOp.MesosUpdate(_, MarathonTaskStatus.Running(mesosStatus), now) if !hasStartedRunning =>
         val updated = copy(
           status = status.copy(
             startedAt = Some(now),
             mesosStatus = mesosStatus))
-        TaskStateChange.Update(updated)
+        TaskStateChange.Update(newState = updated, oldState = Some(this))
 
       // case 2: terminal
-      case TaskStateOp.MesosUpdate(MarathonTaskStatus.Terminal(_), now) =>
-        TaskStateChange.Expunge
+      case TaskStateOp.MesosUpdate(_, MarathonTaskStatus.Terminal(_), now) =>
+        TaskStateChange.Expunge(this)
 
       // case 3: health or state updated
-      case TaskStateOp.MesosUpdate(taskStatus, now) =>
-        val healthOrStateChange = for {
-          current <- status.mesosStatus
-          update <- taskStatus.mesosStatus
-          newStatus <- updatedHealthOrState(current, update)
-        } yield TaskStateChange.Update(copy(status = status.copy(mesosStatus = Some(newStatus))))
-
-        healthOrStateChange.getOrElse{
-          log.debug("Ignoring status update for {}. Status did not change.", taskId)
-          TaskStateChange.NoChange
+      case TaskStateOp.MesosUpdate(_, taskStatus, now) =>
+        updatedHealthOrState(status.mesosStatus, taskStatus.mesosStatus) match {
+          case Some(newStatus) =>
+            val updatedTask = copy(status = status.copy(mesosStatus = Some(newStatus)))
+            TaskStateChange.Update(newState = updatedTask, oldState = Some(this))
+          case None =>
+            log.debug("Ignoring status update for {}. Status did not change.", taskId)
+            TaskStateChange.NoChange(taskId)
         }
 
-      // failure case: Launch
-      case _: TaskStateOp.Launch =>
-        TaskStateChange.Failure("Unable to handle Launch op on LaunchedEphemeral")
+      case TaskStateOp.ForceExpunge(_) =>
+        TaskStateChange.Expunge(this)
 
-      // failure case: Timeout
-      case TaskStateOp.ReservationTimeout =>
-        TaskStateChange.Failure("Unable to handle Timeout op on LaunchedEphemeral")
+      case TaskStateOp.LaunchEphemeral(task) =>
```

Comment:
Should not be called.

I wonder if we could make it more clear that LaunchEphemeral/Reserve/Revert should never be called on existing tasks.

Maybe subdivide TaskStateOp into two sub sealed traits: NewTaskStateOp, ChangeTaskStateOp. Then implement Task.update only for ChangeTaskStateOp?